File: lib/data/fluxpattern.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/data/fluxmap.h"
#include "lib/data/fluxpattern.h"
#include "lib/config/proto.h"
#include "protocol.h"
#include <numeric>
#include <math.h>
#include <strings.h>

FluxPattern::FluxPattern(unsigned bits, uint64_t pattern): _bits(bits)
{
    const uint64_t TOPBIT = 1ULL << 63;

    assert(pattern != 0);

    unsigned lowbit = findLowestSetBit(pattern) - 1;

    pattern <<= 64 - bits;
    _highzeroes = 0;
    while (!(pattern & TOPBIT))
    {
        pattern <<= 1;
        _highzeroes++;
    }

    _length = 0;
    while (pattern != TOPBIT)
    {
        unsigned interval = 0;
        do
        {
            pattern <<= 1;
            interval++;
        } while (!(pattern & TOPBIT));
        _intervals.push_back(interval);
        _length += interval;
    }

    if (lowbit)
    {
        _lowzero = true;
        /* Note that length does *not* include this interval. */
        _intervals.push_back(lowbit + 1);
    }
}

bool FluxPattern::matches(const unsigned* end, FluxMatch& match) const
{
    const double clockDecodeThreshold =
        globalConfig()->decoder().bit_error_threshold();
    const unsigned* start = end - _intervals.size();
    unsigned candidatelength = std::accumulate(start, end - _lowzero, 0);
    if (!candidatelength)
        return false;
    match.clock = (double)candidatelength / (double)_length;

    unsigned exactIntervals = _intervals.size() - _lowzero;
    for (unsigned i = 0; i < exactIntervals; i++)
    {
        double ii = match.clock * (double)_intervals[i];
        double ci = (double)start[i];
        double error = fabs((ii - ci) / match.clock);
        if (error > clockDecodeThreshold)
            return false;
    }

    if (_lowzero)
    {
        double ii = match.clock * (double)_intervals[exactIntervals];
        double ci = (double)start[exactIntervals];
        double error = (ii - ci) / match.clock;
        if (error > clockDecodeThreshold)
            return false;
    }

    match.matcher = this;
    match.intervals = _intervals.size();
    match.zeroes = _highzeroes;
    return true;
}

FluxMatchers::FluxMatchers(
    const std::initializer_list<const FluxMatcher*> matchers):
    _matchers(matchers)
{
    _intervals = 0;
    for (const auto* matcher : matchers)
        _intervals = std::max(_intervals, matcher->intervals());
}

bool FluxMatchers::matches(const unsigned* intervals, FluxMatch& match) const
{
    for (const auto* matcher : _matchers)
    {
        if (matcher->matches(intervals, match))
            return true;
    }
    return false;
}

File: java/com/cowlark/fluxengine/data/FluxPattern.java
package com.cowlark.fluxengine.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * A single flux pattern, ported from lib/data/fluxpattern.{h,cc}.
 */
public class FluxPattern implements FluxMatcher
{
    private static final long TOPBIT = 1L << 63;

    private final int bitCount;
    private final int highZeroes;
    private final List<Integer> intervals = new ArrayList<>();
    private final int length;
    private boolean lowZero = false;

    public FluxPattern(int bits, long pattern)
    {
        bitCount = bits;
        if (pattern == 0)
            throw new IllegalArgumentException("flux pattern may not be zero");
        if (bits < 1 || bits > 64)
            throw new IllegalArgumentException("flux pattern bit count must be 1..64");

        int lowBit = findLowestSetBit(pattern) - 1;

        pattern <<= (64 - bits);
        int highZeroesLocal = 0;
        while ((pattern & TOPBIT) == 0)
        {
            pattern <<= 1;
            highZeroesLocal++;
        }
        highZeroes = highZeroesLocal;

        int lengthLocal = 0;
        while (pattern != TOPBIT)
        {
            int interval = 0;
            do
            {
                pattern <<= 1;
                interval++;
            } while ((pattern & TOPBIT) == 0);
            intervals.add(interval);
            lengthLocal += interval;
        }
        length = lengthLocal;

        if (lowBit != 0)
        {
            lowZero = true;
            intervals.add(lowBit + 1);
        }
    }

    /* Returns the index (1-based) of the lowest set bit, or 0 if none. */
    private static int findLowestSetBit(long value)
    {
        if (value == 0)
            return 0;
        int bit = 1;
        while ((value & 1) == 0)
        {
            value >>= 1;
            bit++;
        }
        return bit;
    }

    @Override
    /* The `endIndex` is one past the newest candidate interval, mirroring the
     * C++ pointer passed as `&*candidates.end()`. */
    public boolean matches(long[] candidates, int endIndex, double clockDecodeThreshold,
            FluxMatch match)
    {
        int start = endIndex - intervals.size();

        int candidateLength = 0;
        for (int i = start; i < endIndex - (lowZero ? 1 : 0); i++)
            candidateLength += candidates[i];

        if (candidateLength == 0)
            return false;
        match.clock = (double) candidateLength / (double) length;

        int exactIntervals = intervals.size() - (lowZero ? 1 : 0);
        for (int i = 0; i < exactIntervals; i++)
        {
            double ii = match.clock * intervals.get(i);
            double ci = candidates[start + i];

            double error = Math.abs((ii - ci) / match.clock);
            if (error > clockDecodeThreshold)
                return false;
        }

        if (lowZero)
        {
            double ii = match.clock * intervals.get(exactIntervals);
            double ci = candidates[start + exactIntervals];

            double error = (ii - ci) / match.clock;
            if (error > clockDecodeThreshold)
                return false;
        }

        match.matcher = this;
        match.intervals = intervals.size();
        match.zeroes = highZeroes;
        return true;
    }

    @Override
    public int intervals()
    {
        return intervals.size();
    }

    /* Package-private accessors for the tests (mirrors the C++ `friend`
     * test_patternconstruction/test_patternmatching). */

    int getBitCount()
    {
        return bitCount;
    }

    List<Integer> getIntervals()
    {
        return Collections.unmodifiableList(intervals);
    }

    int getHighZeroes()
    {
        return highZeroes;
    }
}